dipy/reconst/dki.py

```diff
@@ -700,8 +800,427 @@ def split_dki_param(dki_params):
+    min_separation_angle : float in [0, 90], optinal
+        The minimum distance between directions. If two peaks are too close
+        only the larger of the two is returned.
+    mask : array, optional
+        If `mask` is provided, voxels that are False in `mask` are skipped and
+        no peaks are returned.
+    return_odf : bool, optional
+        If True, the odfs sampled on sphere directions are returned.
+    normalize_peaks : bool, optional
+        If true, all peak values are calculated relative to `max(odf)`.
+    npeaks : int, optional
+        Maximum number of peaks found (default 3 peaks).
+    gtol : float, optional
+        Degree gradient must be less than gtol before succesful termination. If
+        gtol is None, fiber direction is directly taken from the initial
+        sampled directions of the given sphere object
+
+    Returns
+    -------
+    pam : PeaksAndMetrics
+        An object with ``peak_directions``, ``peak_values``, ``peak_indices``,
+        ``odf`` as attributes
+
+    .. [Jen2014] Jensen, J.H., Helpern, J.A., Tabesh, A., (2014). Leading
+        non-Gaussian corrections for diffusion orientation distribution
+        function. NMR Biomed. 27, 202-211. http://dx.doi.org/10.1002/nbm.3053.
+
+    .. [Raf2015] Neto Henriques, R., Correia, M.M., Nunes, R.G., Ferreira,
+        H.A. (2015). Exploring the 3D geometry of the diffusion kurtosis
+        tensor - Impact on the development of robust tractography procedures
+        and novel biomarkers. NeuroImage 111: 85-99.
+        doi:10.1016/j.neuroimage.2015.02.004
+    """
+    shape = dki_params.shape[:-1]
+
+    # select voxels where to find fiber directions
+    if mask is None:
+        mask = np.ones(shape, dtype='bool')
+    else:
+        if mask.shape != shape:
+            raise ValueError("Mask is not the same shape as data.")
+    evals, evecs, kt = split_dki_param(dki_params)
+    # select non-zero voxels
+    # (when #677 is merged replace this code lines by function _positive_evals)
+    # pos_evals = _positive_evals(evals[..., 0], evals[..., 1], evals[..., 2])
+    er = np.finfo(evals.ravel()[0]).eps * 1e3
+    pos_evals = np.logical_and(evals[..., 0] > er,
+                               np.logical_and(evals[..., 1] > er,
+                                              evals[..., 2] > er))
+    mask = np.logical_and(mask, pos_evals)
+
+    # initialize parameters for pam
+    qa_array = np.zeros((shape + (npeaks,)))
+    peak_dirs = np.zeros((shape + (npeaks, 3)))
+    peak_values = np.zeros((shape + (npeaks,)))
+    peak_indices = np.zeros((shape + (npeaks,)), dtype='int')
+    if return_odf:
+        odf_array = np.zeros((shape + (len(sphere.vertices),)))
+
+    # for all voxels:
+    global_max = -np.inf
+    for idx in ndindex(shape):
+        if not mask[idx]:
+            continue
+        
+        # Compute dimensionless tensor U
+        dt = np.dot(np.dot(evecs[idx], np.diag(evals[idx])), evecs[idx].T)
```

Comment:
Maybe compute this outside of the loop? You could create a `U` array upfront for all voxels in one move.